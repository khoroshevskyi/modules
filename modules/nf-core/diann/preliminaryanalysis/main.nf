process DIANN_PRELIMINARYANALYSIS {
    tag "$ms_file.baseName"
    label 'process_high'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://containers.biocontainers.pro/s3/SingImgsRepo/diann/v1.8.1_cv1/diann_v1.8.1_cv1.img' :
        'docker.io/biocontainers/diann:v1.8.1_cv1' }"

    input:
    tuple val(meta), path(ms_file), path(predict_library)

    output:
    tuple val(meta), path("*.quant"), emit: diann_quant
    tuple val(meta), path("*_diann.log"), emit: log
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "DIANN_PRELIMINARYANALYSIS module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    def args = task.ext.args ?: ''

    """
    diann   --lib ${predict_library} \\
            --f ${ms_file} \\
            --threads ${task.cpus} \\
            --temp ./ \\
            ${args}

    cp report.log.txt ${ms_file.baseName}_diann.log

    # Check for mzML indexing errors first (specific format issue)
    if grep -q -E "(No index list offset found|Failure reading input file)" ${ms_file.baseName}_diann.log; then
        echo "ERROR: Invalid mzML format - missing index offset or corrupted file"
        echo "DIA-NN requires fully indexed mzML files with proper indexListOffset and valid MS2 spectra"
        exit 1
    fi

    # Check for other serious errors
    if grep -q -E "(ERROR.*failed to load.*files:|ERROR.*cannot load the file.*skipping|No MS2 spectra.*aborting)" ${ms_file.baseName}_diann.log; then
        echo "ERROR: DIA-NN PRELIMINARY ANALYSIS failed - check log for file loading errors"
        exit 1
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        DIA-NN: \$(diann 2>&1 | grep "DIA-NN" | grep -oP "\\d+\\.\\d+(\\.\\w+)*(\\.[\\d]+)?")
    END_VERSIONS
    """

    stub:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "DIANN_PRELIMINARYANALYSIS module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    def args = task.ext.args ?: ''

    """
    touch ${ms_file.baseName}.quant
    touch ${ms_file.baseName}_diann.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        DIA-NN: \$(diann 2>&1 | grep "DIA-NN" | grep -oP "\\d+\\.\\d+(\\.\\w+)*(\\.[\\d]+)?")
    END_VERSIONS
    """
}
