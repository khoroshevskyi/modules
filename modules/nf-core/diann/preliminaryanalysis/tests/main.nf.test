nextflow_process {

    name "Test Process DIANN_PRELIMINARYANALYSIS"
    script "../main.nf"
    process "DIANN_PRELIMINARYANALYSIS"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "diann"
    tag "diann/preliminaryanalysis"
    tag "untar"

    setup {
        run("UNTAR") {
            script "../../../../nf-core/untar/main.nf"
            process {
                """
                input[0] = [
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'proteomics/diann/RD139_Narrow_UPS1_0_1fmol_inj1.mzML.tar.gz', checkIfExists: true)
                ]
                """
            }
        }
    }

    test("proteomics - preliminary analysis") {

        when {
            process {
                """
                input[0] = UNTAR.out.untar.map{ meta, dir ->  
                    def mzmlFile = dir.resolve('RD139_Narrow_UPS1_0_1fmol_inj1.mzML')
                    [meta + [precursormasstolerance:'20', fragmentmasstolerance:'20', precursormasstoleranceunit:'ppm', fragmentmasstoleranceunit:'ppm'], mzmlFile, file(params.modules_testdata_base_path + 'proteomics/diann/lib.predicted.speclib', checkIfExists: true)]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert path("${process.out.log[0][1]}").exists() },
                { assert snapshot(
                    process.out.diann_quant,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("proteomics - preliminary analysis - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = UNTAR.out.untar.map{ meta, dir ->  
                    def mzmlFile = dir.resolve('RD139_Narrow_UPS1_0_1fmol_inj1.mzML')
                    [meta + [precursormasstolerance:'20', fragmentmasstolerance:'20', precursormasstoleranceunit:'ppm', fragmentmasstoleranceunit:'ppm'], mzmlFile, file(params.modules_testdata_base_path + 'proteomics/diann/lib.predicted.speclib', checkIfExists: true)]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
