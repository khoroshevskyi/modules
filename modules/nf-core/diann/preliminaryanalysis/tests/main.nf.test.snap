{
    "proteomics - preliminary analysis - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "precursormasstolerance": "20",
                            "fragmentmasstolerance": "20",
                            "precursormasstoleranceunit": "ppm",
                            "fragmentmasstoleranceunit": "ppm"
                        },
                        "peakpicker_tutorial_1.quant:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "precursormasstolerance": "20",
                            "fragmentmasstolerance": "20",
                            "precursormasstoleranceunit": "ppm",
                            "fragmentmasstoleranceunit": "ppm"
                        },
                        "peakpicker_tutorial_1_diann.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,97f6ef15723946b0ef56bcbc8762fe1c"
                ],
                "diann_quant": [
                    [
                        {
                            "id": "test",
                            "precursormasstolerance": "20",
                            "fragmentmasstolerance": "20",
                            "precursormasstoleranceunit": "ppm",
                            "fragmentmasstoleranceunit": "ppm"
                        },
                        "peakpicker_tutorial_1.quant:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "log": [
                    [
                        {
                            "id": "test",
                            "precursormasstolerance": "20",
                            "fragmentmasstolerance": "20",
                            "precursormasstoleranceunit": "ppm",
                            "fragmentmasstoleranceunit": "ppm"
                        },
                        "peakpicker_tutorial_1_diann.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,97f6ef15723946b0ef56bcbc8762fe1c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-24T13:46:59.149221"
    },
    "proteomics - preliminary analysis": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "precursormasstolerance": "20",
                        "fragmentmasstolerance": "20",
                        "precursormasstoleranceunit": "ppm",
                        "fragmentmasstoleranceunit": "ppm"
                    },
                    "peakpicker_tutorial_1_mzML.quant:md5,24b3cf360ef26805299bc80051e42ff2"
                ]
            ],
            [
                "versions.yml:md5,97f6ef15723946b0ef56bcbc8762fe1c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-24T13:49:12.996339"
    }
}